docs: IA Phase 2 (P0/P1): add Connect "Source control and access scope" page (#723)
Promptless/docs: src/content/docs/docs/connect/source-control.mdx

Reviewer: Overall, prioritize GitLab over BitBucket. Priority in ordering should generally be GitHub, GitLab, then BitBucket.

---
title: Source control and access scope
description: Understand how Promptless connects to GitHub, GitHub Enterprise, Bitbucket, and GitLab, what access each connection grants, and how to keep that access least-privilege.
slug: docs/connect/source-control
sidebar:
  hidden: false
  order: 25
  label: Source control
---
import { Aside } from '@astrojs/starlight/components';

> Source control is where Promptless watches for changes and publishes documentation. A single Git provider connection can do up to three jobs — firing triggers, supplying code context, and opening documentation pull requests — and each job needs a different level of access. This page covers what each provider connection does, the access it grants, and how to keep that access scoped to the minimum Promptless needs.

Reviewer: ```suggestion
Source control is where Promptless watches for changes and publishes documentation. A single Git provider connection can do up to three jobs, each of which needs a different level of access: firing triggers, supplying code context, and opening documentation pull requests. This page covers what each provider connection does, the access it grants, and how to keep that access scoped to the minimum Promptless needs.
```

| Provider                                                            | Used for                                            |
| ------------------------------------------------------------------- | --------------------------------------------------- |
| [GitHub](/docs/reference/integrations/github)                       | Triggers, context, and publishing                   |
| [GitHub Enterprise](/docs/reference/integrations/github-enterprise) | Triggers, context, and publishing (Server or Cloud) |
| [Bitbucket](/docs/reference/integrations/bitbucket)                 | Triggers                                            |
| [GitLab](/docs/reference/integrations/gitlab)                       | Triggers and context                                |

Only GitHub-based collections receive documentation pull requests. Bitbucket and GitLab connect as trigger sources (and, for GitLab, context sources): Promptless monitors their pull and merge requests but does not publish documentation to them.

## How each connection authenticates

Each provider uses its own authentication model, and each one limits Promptless to the repositories and actions you authorize.

### GitHub

> Promptless connects through its official GitHub App and authenticates with JWTs generated by the app installation. The app grants read and write access to the repositories you select, and either you or Promptless can revoke that access at any time. During installation, you choose which repositories to grant — typically your source code and, if your docs live in GitHub, your documentation repo.

Reviewer: ```suggestion
Promptless connects through its official GitHub App and authenticates with JWTs generated by the app installation. The app grants read and write access to the repositories you select, and either you or Promptless can revoke that access at any time. During installation, you choose which repositories to grant. These are typically your source code and, if your docs live in GitHub, your documentation repo.
```

### GitLab

GitLab connects with a group access token that carries the `api` scope. Group access tokens have read-write `api` scope by default, but Promptless performs only read operations with the token — viewing merge requests, reading repository files, and fetching project metadata. Promptless never pushes code, creates merge requests, or modifies your GitLab projects through this token.

<Aside type="tip">
  If your organization requires stricter access controls, provision the token with the `read_api` scope instead of `api`. That limits the token to read-only operations at the GitLab API level, adding a layer of security beyond Promptless's own behavioral restrictions.
</Aside>

## Scope publishing access to your docs

When you connect GitHub for publishing, the Promptless GitHub App has write access to your documentation repository so it can open pull requests with suggested changes. Use **path scope** to bound where those writes land.

> Path scope controls which files Promptless can modify in a documentation repository. Specify directories during doc collection setup, and Promptless enforces those boundaries: with no directories set, it can modify files anywhere in the repo; with directories set, it can modify only files that match one of those paths or nest inside them. Path scope accepts both directories and individual files, so an entry like `docs/` allows any file under that directory, while `CHANGELOG.md` allows only that one file. If Promptless drafts a change outside the configured scope, it rejects the suggestion and returns an error listing the out-of-scope files — useful for monorepos where docs live alongside source code.

Reviewer: ```suggestion
Path scope controls which files Promptless can modify in a documentation repository. Specify directories during doc collection setup, and Promptless enforces those boundaries: with no directories set, it can modify files anywhere in the repo, but with directories set, it can modify only files that match one of those paths or nest inside them. Path scope accepts both directories and individual files, so an entry like `docs/` allows any file under that directory, while `CHANGELOG.md` allows only that one file. If Promptless drafts a change outside the configured scope, it rejects the suggestion and returns an error listing the out-of-scope files, which is useful for monorepos where docs live alongside source code.
```

For the full doc collection setup, including path scope and platform support, see [GitHub repos (docs as code)](/docs/connect/doc-locations/github-repos).

## Keep access least-privilege

> Promptless follows the principle of least privilege, granting only the access each job needs. Its connections give you several ways to hold that line:

Reviewer: ```suggestion
Promptless follows the principle of least privilege, granting only the access each job needs. Its connections give you several ways to enforce that:
```

- **Select only the repositories Promptless needs.** On GitHub, choose **Only select repositories** rather than **All repositories**, and manage the list from your GitHub organization settings.
- **Prefer read-only access where you don't publish.** On GitLab, a `read_api` token keeps the connection read-only. For a public open-source repository, the [read-only GitHub app](/docs/start-here/open-source-quickstart) opens documentation pull requests from a fork instead of taking write access to your repo.
- **Revoke access at any time.** You or Promptless can revoke the GitHub App connection, and you can rotate or remove provider tokens (Bitbucket app password, GitLab group access token) on the provider side.

> For how source-control access fits Promptless's broader security posture — what it reads, what it retains, and how tenants are isolated — see [Network architecture](/docs/security/network-architecture) and [Data handling and classification](/docs/security/data-handling-and-classification).

Reviewer: ```suggestion
For how source-control access fits Promptless's broader security posture, see [Network architecture](/docs/security/network-architecture) and [Data handling and classification](/docs/security/data-handling-and-classification).
```
